Source/WebKit/UIProcess/RemoteLayerTree/RemoteAcceleratedEffectStack.mm

```diff
@@ -35,94 +35,137 @@
-    : m_acceleratedTimelineTimeOrigin(acceleratedTimelineTimeOrigin)
+RemoteAcceleratedEffectStack::RemoteAcceleratedEffectStack(WebCore::FloatRect bounds, Seconds acceleratedTimelineTimeOrigin)
+    : m_bounds(bounds)
+    , m_acceleratedTimelineTimeOrigin(acceleratedTimelineTimeOrigin)
 {
 }
 
 void RemoteAcceleratedEffectStack::setEffects(AcceleratedEffects&& effects)
 {
     AcceleratedEffectStack::setEffects(WTFMove(effects));
 
+    bool affectsOpacity = false;
+    bool affectsTransform = false;
+
     for (auto& effect : m_primaryLayerEffects) {
-        if (effect->animatedProperties().contains(AcceleratedEffectProperty::Opacity)) {
-            m_affectedLayerProperties.add(LayerProperty::Opacity);
-            return;
-        }
+        auto& properties = effect->animatedProperties();
+        affectsOpacity = affectsOpacity || properties.contains(AcceleratedEffectProperty::Opacity);
+        affectsTransform = affectsTransform || properties.containsAny(transformRelatedAcceleratedProperties);
+        if (affectsOpacity && affectsTransform)
+            break;
     }
+
+    if (affectsOpacity)
+        m_affectedLayerProperties.add(LayerProperty::Opacity);
+    if (affectsTransform)
+        m_affectedLayerProperties.add(LayerProperty::Transform);
 }
 
 #if PLATFORM(MAC)
 void RemoteAcceleratedEffectStack::initEffectsFromMainThread(PlatformLayer *layer, MonotonicTime now)
 {
     ASSERT(!m_opacityPresentationModifier);
+    ASSERT(!m_transformPresentationModifier);
     ASSERT(!m_presentationModifierGroup);
 
-    if (!m_affectedLayerProperties.contains(LayerProperty::Opacity))
+    if (!m_affectedLayerProperties.containsAny({ LayerProperty::Opacity, LayerProperty::Transform }))
         return;
 
     auto computedValues = computeValues(now);
-    auto *opacity = @(computedValues.opacity);
 
-    m_presentationModifierGroup = [CAPresentationModifierGroup groupWithCapacity:1];
-    m_opacityPresentationModifier = adoptNS([[CAPresentationModifier alloc] initWithKeyPath:@"opacity" initialValue:opacity additive:NO group:m_presentationModifierGroup.get()]);
+    auto numberOfPresentationModifiers = m_affectedLayerProperties.containsAll({ LayerProperty::Opacity, LayerProperty::Transform }) ? 2 : 1;
+    m_presentationModifierGroup = [CAPresentationModifierGroup groupWithCapacity:numberOfPresentationModifiers];
 
-    [layer addPresentationModifier:m_opacityPresentationModifier.get()];
+    if (m_affectedLayerProperties.contains(LayerProperty::Opacity)) {
+        auto *opacity = @(computedValues.opacity);
+        m_opacityPresentationModifier = adoptNS([[CAPresentationModifier alloc] initWithKeyPath:@"opacity" initialValue:opacity additive:NO group:m_presentationModifierGroup.get()]);
+        [layer addPresentationModifier:m_opacityPresentationModifier.get()];
+    }
+
+    if (m_affectedLayerProperties.contains(LayerProperty::Transform)) {
```

Comment:
How are you taking care of the geometry math here? You need to account for `transform-origin`,  and recall that layer size is often not equal to border box size because of shadows, children etc.

Reply:
Per my testing, the `transform-origin` is already handled by setting the appropriate `CALayer` properties statically. Our layout tests hasn't found an issue with the geometry either. I will be doing some more manual testing in this area in the coming weeks to double check on that.